src/OrchardCore.Modules/OrchardCore.Media/Filters/MediaFilters.cs

```diff
@@ -76,4 +79,35 @@ public ValueTask<FluidValue> ProcessAsync(FluidValue input, FilterArguments argu
             return new ValueTask<FluidValue>(new StringValue(url));
         }
     }
+
+    public class AppendVersionFilter : ILiquidFilter
+    {
+        private readonly IFileVersionProvider _fileVersionProvider;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+
+        private string _pathBase;
+
+        public AppendVersionFilter(
+            IFileVersionProvider fileVersionProvider,
+            IHttpContextAccessor httpContextAccessor)
+        {
+            _fileVersionProvider = fileVersionProvider;
+            _httpContextAccessor = httpContextAccessor;
+        }
+
+        public ValueTask<FluidValue> ProcessAsync(FluidValue input, FilterArguments arguments, TemplateContext ctx)
+        {
+            var url = input.ToStringValue();
+
+            if (String.IsNullOrEmpty(_pathBase))
+            {
+                //for media we must append Startup.AssetsUrlPrefix to pathBase or IFileVersionProvider will not find it
+                _pathBase = String.Concat(_httpContextAccessor.HttpContext.Request.PathBase.ToString(), Startup.AssetsUrlPrefix);
```

Comment:
Hmm, with only 2 strings to concat, maybe just.

    _pathBase = _httpContextAccessor.HttpContext.Request.PathBase + Startup.AssetsUrlPrefix;

Yes, we need to add this prefix to the `PathBase` because the `RequestPath` of the related static middleware is set with it, so the generated url need to include the `/media` segment. If i use this filter before the `asset_url` we don't need this. But normal to use it after. Here we do a kind of mapping, so just for testing i tried to use the `PathBase` as it is, use a `MapPublicUrlToPath` before `AddFileVersionToPath` and then `MapPathToPublicUrl` and it was working.

 What i also tried is to not set the `RequestPath` of the middleware, then init the `MediaFileStore` with a `mediaUrlBase` not including `/media`, just the tenant url prefix, but still having the `mediaFileProvider` with a root including `/media`. Then it was working without having to add the asset prefix here, the only difference is that generated urls don't include `/media`.

If we can do this we could have a more generic filter, not specific to media. No need to append the prefix to the `PathBase`. Note: In this case, for the default tenant `_pathBase = Request.PathBase.ToString()` would be always empty, so we would have to check `if (_pathBase == null)`, but here no more need to cache it.

 But maybe intended to require this segment, hmm the middleware is selected faster, but if other static middlewares run before, i don't know. So, this is just to share my findings, and because we use another way to append versions for our style / script tags, maybe not so important to not have a more generic filter.

Hmm, here we rely on the `OC.Media.Startup.AssetsUrlPrefix` used to init the media file store which do the mapping, but what happens if this service is overriden but still using the same filter? That's why above i thought about using the media store mappings (but not the best to do 2 mapping), or maybe the media store could expose its `PublicUrlBase` or `MediaUrlBase` (i don't like all these different names) and just use it (blob storage aside ;) where it is currently a full uri) without having to use the request `PathBase + assetPrefix`.

Note: Maybe we don't use enough the `PathString` type.

Reply:
Thanks @jtkech I also thinking of this virtual/relative path issue this morning, and try not using the RequestPath of middleware as well, but that was not so good for other reasons, not least because then caused more path mangling when using in multi-tenant scenario when `HttpContext.PathBase` has a value.

Also the whole thing never worked properly with the mvc img helper as it did not path wrangling.

So I settle on moving logic regarding this to `IFileVersionProvider` and try implementing a `IVirtualPathBaseProvider` on `IMediaFileProvider` to provide base virtual path (also stuck on all these different names!). So that works  when `HttpContext.PathBase` has a value in multi-tenant and when has it no value. And allows overriding of `Media` as the virtual path base is configured during `IMediaFileProvider` resolution, so can be overridden.

This now lets it work properly with mvc tag helper as well.

So much mangling of different possible paths (gotta love virtual paths!)

I have put `IVirtualPathBaseProvider` in `OC.Abstractions` but not sure if this the best place.

I have left the extra cache checks in as while removing them works for our exact scenario, but if another implementation, or other `IFileProvider` registration handles virtual paths and base paths differently there seemed the chance of missing a cache hit. At the moment the watchers work well if I don't have an asset available, then copy it into whichever folder while site is running.

> Note: Maybe we don't use enough the PathString type.
Maybe we don't, but when trying to use it, it required adding references to projects that maybe don't need them...